### PR TITLE
Fix restart code so you can restart with different numbers of MPI ranks

### PR DESCRIPTION
**Pull request by Yurlungur**
As requested by @soumide1102 . This appears to work for tracer particles at least. I've tested (by hand, not formally) that the tracer data is undamaged for a simple test with 2 and 4 MPI ranks. I go from 4, restart 2, run a while, and then restart from 4 again.

The strategy is: I count the number of particle datasets to read from the HDF5 restart file file and then read that many. If there are too few, not all ranks read any particles.  If there are too many, modulo arithmetic is used to spread the particles round-robin across the ranks. 

After reading the datasets, I introduce a new step where I leverage the particle communication machinery. Each rank communicates particles that aren't within it's "bounding box" to one of it's 26 neighbor ranks. (Particle communication crosses edges and corners, unlike fluid variable communication.) This communication step must be performed a large number of times---as many times as it might take for a particle to transition from a rank at one end of the grid to a rank at the opposite end of the grid. I do it `2*(N1CPU+N2CPU+N3CPU)` times, which is almost certainly overkill. This does imply a performance hit when the code restarts. But the operation should be essentially a no-op once all particles have been communicated.

@brryan I'd be interested in a sanity check from you that this makes sense. @soumide1102 please try it out.

### CODE REVIEW

**Comment by brryan** [comment]
Glad this seemed to just work! Your test of this feature is convincing to me, @Yurlungur. Two comments now before I get into it. 1) the fluid should actually communicate corners and edges in the same way as the particles here -- it's the same pattern of Sendrecv calls, actually. IIRC you need to communicate corners for flux CT. 2) Would it always be sufficient to only do communication `MAX(N1CPU, N2CPU, N3CPU)` times? I *think* this should communicate particles even all the way along the diagonal of a cube. Although if the restart cost isn't hugely awful, it's only an initialization cost so probably not worth investigating this.

File: core/io.c [hunk 1]
```diff
@@ -1494,25 +1495,43 @@ void restart_read(char *fname) {
 #undef RANK
   }
   // Superphoton datasets -- each processor must create all datasets
+  // compute how many datasets are needed
+  char dsetnam[STRLEN];
+  int  num_datasets = -1;
```
Comment (brryan) [inline core/io.c:1500]: Nice way of doing this

File: core/io.c [hunk 2]
```diff
@@ -1494,25 +1495,43 @@ void restart_read(char *fname) {
...
+  hid_t * ph_dsets   = safe_malloc((size_t)num_datasets * sizeof(hid_t));
+  size_t *dset_sizes = safe_malloc((size_t)num_datasets * sizeof(size_t));
+  for (int n = 0; n < num_datasets; n++) {
     sprintf(dsetnam, "photons_%08d", n);
     ph_dsets[n] = H5Dopen(file_id, dsetnam, H5P_DEFAULT);
     hid_t space = H5Dget_space(ph_dsets[n]);
     H5Sget_simple_extent_dims(space, dims, NULL);
-    if (n == mpi_myrank())
-      nph_in = dims[0];
+    if (n % mpi_nprocs() == mpi_myrank()) {
```
Comment (brryan) [inline core/io.c:1517]: This is the optimal way of doing this, I think

File: core/io.c [hunk 3]
```diff
@@ -1696,6 +1716,18 @@ int restart_init() {
 #if SCATTERING
       { init_all_hotcross(); }
 #endif
+
+      // Check boundary conditions and put photons where they belong
+      if (mpi_myrank() == 0) {
+        printf("Shuffling and load balancing particles\n");
```
Comment (brryan) [inline core/io.c:1722]: Is this load balancing or just getting all their particles to the MPI rank that owns their spatial locations?

**Review by brryan — APPROVED** [review]
looks good to me!

File: core/io.c (hunk 3 above)
Comment (Yurlungur) [reply core/io.c:1722]: The latter. I will fix the wording.

**Comment by Yurlungur** [comment]
> Glad this seemed to just work! Your test of this feature is convincing to me, @Yurlungur. 

Cool! Thanks for taking a look @brryan !

> 1) the fluid should actually communicate corners and edges in the same way as the particles here -- it's the same pattern of Sendrecv calls, actually. IIRC you need to communicate corners for flux CT.

Oops my bad. Yeah you're right. I was thinking of without B-fields.

2) Would it always be sufficient to only do communication `MAX(N1CPU, N2CPU, N3CPU)` times? I *think* this should communicate particles even all the way along the diagonal of a cube.

Hmm I think you're right. I was being very conservative in my choice. The max should work because we're communicating along corners.

**Comment by Yurlungur** [comment]
I'm going to merge this in. Behaviour should be unchanged if the number of ranks doesn't change between runs.

**Pull request merged**